### PR TITLE
[openacc][openmp] Add dialect representation for acc atomic operations

### PR DESCRIPTION
**Pull request by razvanlupusoru**
The OpenACC standard specifies an `atomic` construct in section 2.12 (of 3.3 spec), used to ensure that a specific location is accessed or updated atomically. Four different clauses are allowed: `read`, `write`, `update`, or `capture`. If no clause appears, it is as if `update` is used.

The OpenMP specification defines the same clauses for `omp atomic`. The types of expression and the clauses in the OpenACC spec match the OpenMP spec exactly. The main difference is that the OpenMP specification is a superset - it includes clauses for `hint` and `memory order`. It also allows conditional expression statements. But otherwise, the expression definition matches.

Thus, for OpenACC, we refactor and reuse the OpenMP implementation as follows:
* The atomic operations are duplicated in OpenACC dialect. This is preferable so that each language's semantics are precisely represented even if specs have divergence.
* However, since semantics overlap, a common interface between the atomic operations is being added. The semantics for the interfaces are not generic enough to be used outside of OpenACC and OpenMP, and thus new folders were added to hold common pieces of the two dialects.
* The atomic interfaces define common accessors (such as getting `x` or `v`) which match the OpenMP and OpenACC specs. It also adds common verifiers intended to be called by each dialect's operation verifier.
* The OpenMP write operation was updated to use `x` and `expr` to be consistent with its other operations (that use naming based on spec).

The frontend lowering necessary to generate the dialect can also be reused. This will be done in a follow up change.

Differential Revision: https://reviews.llvm.org/D158768

### CODE REVIEW

File: mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td [hunk 1]
```diff
@@ -1207,6 +1209,153 @@ def OpenACC_YieldOp : OpenACC_Op<"yield", [ReturnLike, Terminator,
...
+  let hasRegionVerifier = 1;
+  let hasCanonicalizeMethod = 1;
+  let extraClassDeclaration = [{
+    Operation* getFirstOp() {
+      return &getRegion().front().getOperations().front();
+    }
+  }];
+}
+
+def AtomicCaptureOp : OpenACC_Op<"atomic.capture",
```
Comment (vzakhari) [inline mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td:1308]: I wonder if you want to make it `RecursiveMemoryEffects`.

File: mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    is atomically written/read.
+    
+    Since atomic update expression comes in many forms, this interface requires
+    that the operation uses a region with a single argument to capture the
+    expression.
+
+    The region describes how to update the value of `x`. It takes the value at
+    `x` as an input and must yield the updated value. Only the update to `x` is
+    atomic. Generally the region must have only one instruction, but can
+    potentially have more than one instructions too. The update is sematically
```
Comment (vzakhari) [inline mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td:129]:
```suggestion
    potentially have more than one instructions too. The update is semantically
```

File: mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    InterfaceMethod<[{
+        Obtains `x` which is the address to which the value is atomically
+        written to / read from.
+      }],
+      /*retTy=*/"::mlir::Value",
+      /*methodName=*/"getX",
+      /*args=*/(ins)
+    >,
+    InterfaceMethod<[{
+        Returns true if the new value is same as old value and the operation is
```
Comment (vzakhari) [inline mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td:144]: The comment seems unrelated to the method.

**Review by vzakhari — COMMENTED** [review]
Looks great! Thank you for the changes!  I have just minor remarks.

File: mlir/include/mlir/Dialect/OpenACCMPCommon/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+add_subdirectory(Interfaces)
```
Comment (clementval) [inline mlir/include/mlir/Dialect/OpenACCMPCommon/CMakeLists.txt:1]: Missing new line.

File: mlir/lib/Dialect/OpenACCMPCommon/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+add_subdirectory(Interfaces)
```
Comment (clementval) [inline mlir/lib/Dialect/OpenACCMPCommon/CMakeLists.txt:1]: Missing new line

File: mlir/lib/Dialect/OpenACCMPCommon/Interfaces/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+ADDITIONAL_HEADER_DIRS
+${MLIR_MAIN_INCLUDE_DIR}/mlir/Dialect/OpenACCMPCommon/Interfaces
+
+DEPENDS
+MLIRAtomicInterfacesIncGen
+
+LINK_LIBS PUBLIC
+MLIRIR
+)
```
Comment (clementval) [inline mlir/lib/Dialect/OpenACCMPCommon/Interfaces/CMakeLists.txt:12]: New line

**Review by clementval — COMMENTED** [review]
Looks ok for me.

File: mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td (hunk 3 above)
Comment (razvanlupusoru) [reply mlir/include/mlir/Dialect/OpenACCMPCommon/Interfaces/AtomicInterfaces.td:144]: Great catch.

File: mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td (hunk 1 above)
Comment (razvanlupusoru) [reply mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td:1308]: I agree - and actually I will make the change for both OpenMP and OpenACC.

File: mlir/test/Dialect/OpenMP/invalid.mlir [hunk 7]
```diff
@@ -874,7 +874,7 @@ func.func @omp_atomic_update(%x: memref<i32>, %expr: i32) {
 // -----
 
 func.func @omp_atomic_capture(%x: memref<i32>, %v: memref<i32>, %expr: i32) {
-  // expected-error @below {{expected three operations in omp.atomic.capture region}}
+  // expected-error @below {{expected three operations in atomic.capture region}}
```
Comment (shraiysh) [inline mlir/test/Dialect/OpenMP/invalid.mlir:877]: nit: please add testcases in `OpenACC/invalid.mlir`.

**Review by shraiysh — COMMENTED** [review]
This looks great! Thank you for working on this.

File: mlir/test/Dialect/OpenMP/invalid.mlir (hunk 7 above)
Comment (razvanlupusoru) [reply mlir/test/Dialect/OpenMP/invalid.mlir:877]: Done :)

File: mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td [hunk 8]
```diff
@@ -1207,6 +1209,153 @@ def OpenACC_YieldOp : OpenACC_Op<"yield", [ReturnLike, Terminator,
...
+    /// Otherwise, it returns nullptr.
+    AtomicReadOp getAtomicReadOp();
+
+    /// Returns the `atomic.write` operation inside the region, if any.
+    /// Otherwise, it returns nullptr.
+    AtomicWriteOp getAtomicWriteOp();
+
+    /// Returns the `atomic.update` operation inside the region, if any.
+    /// Otherwise, it returns nullptr.
+    AtomicUpdateOp getAtomicUpdateOp();
```
[the remark this answers is not in the document]
Comment (razvanlupusoru) [reply mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td:1355]: The reason I did not make them part of the interface is that the two dialects have separate operations for the atomics. Namely there is mlir::acc::AtomicUpdateOp and mlir::omp::AtomicUpdateOp.
I could make them part of the same interface as long as we return an interface instead. Something like:
AtomicUpdateOpInterface getAtomicUpdateOp();

What do you think?

[the remark this answers is not in the document]
Comment (shraiysh) [reply mlir/include/mlir/Dialect/OpenACC/OpenACCOps.td:1355]: Good point. I did not observe that. The current version is also okay for me.

**Review by shraiysh — APPROVED** [review]
Looks good to me!

**Review by vzakhari — APPROVED** [review]

**Review by clementval — APPROVED** [review]

**Comment by razvanlupusoru** [comment]
@aartbik Since I added new interfaces, several cmake files had to be updated. I attempted to do the same for the bazel build but did not test it. Would you mind looking through the changes? Thank you!

**Comment by aartbik** [comment]
> @aartbik Since I added new interfaces, several cmake files had to be updated. I attempted to do the same for the bazel build but did not test it. Would you mind looking through the changes? Thank you!

It is very kind of you to include the bazel builds too (especially since this is not a requirement for submitting to MLIR), since it helps people integrating changes into bazel-based repos enormously.  The changes look good to me (although indeed actually testing them will reveal any lingering issue) and will help integrators a lot!

**Comment by grypp** [comment]
Hey, I've just seen this PR, I understand the need for a common atomic interface between these two models. 

I am little concerned about the `OpenACCMPCommon` folder is in the `dialect` folder, but there is no dialect definition here? There is a single interface `AtomicInterface`, imho atomics can be used beyond just OpenACC/MP. 

Would make sense to delete the `OpenACCMPCommon` folder and move the `AtomicInterface` to the `include/mlir/interfaces`?

**Comment by grypp** [comment]
Hey Razvan - long time no chat :) It's great that you are improving the atomic support, nice progress!

I trust your judgment that the interface being OpenMP/OpenACC specific. In that case, would it be a good idea creating `include/mlir/interfaces/OpenACCMPCommonInterfaces`? You might even need more interfaces between these models, so the folder can be useful. (or anything doesn't involve `Dialect` folder works for me.) What do you think?

I also gave a thought of placing the interface in either the OpenMP or OpenACC folder. But I completely agree with you that it might not be the best approach.

**Comment by razvanlupusoru** [comment]
> You might even need more interfaces between these models
Agreed! There will definitely be more interfaces! I also imagine not just operation interfaces - but type interfaces too.

> In that case, would it be a good idea creating include/mlir/interfaces/OpenACCMPCommonInterfaces
Your suggestion makes sense! My main hesitation is just that it feels eventually we may share more than just interfaces. For example, utilities and ops. I don't want to move things around often - and the current home seems it can be a good home even for future expansion.
